### DISCUSSION TITLE
Plugins - what to use in tox 4 instead of tox_cleanup hook?

### DISCUSSION THREAD

**Discussion by f3flight** [tox-dev/tox#2974] [Q&A] [answered]
tox_cleanup is gone and I don't see an appropriate alternative. I need a hook which runs after *all* environments were run (i.e. right before tox shutdown).

More specifically, here's some of the code I used in tox_cleanup:
```
    sdist = tarfile.open(session.package)
    pkg_info = sdist.getmember([name for name in sdist.getnames() if Path(name).name == 'PKG-INFO'][0])
    pkg_info_content = sdist.extractfile(pkg_info).readlines()
    package_name = [line.split(b' ')[1] for line in pkg_info_content if line.startswith(b'Name:')][0].strip().decode()
    package_version = [line.split(b' ')[1] for line in pkg_info_content if line.startswith(b'Version:')][0].strip().decode()
```
Basically I need to figure out the main package true name and version, and then I do some stuff with this after this code, and this runs once per run at the very end. Can I do the same in tox 4?

**Comment by gaborbernat** [comment]
Can you please clarify what exactly you need to do? Beyond just something.

**Reply by f3flight** [reply]
Internal business logic for CI/CD integration, i.e. read some files, write some files, exact actions not really relevant, you can just assume it's some custom code or script not related to tox.
The requirement is to a) know the name and version of the package tox processed and b) preferably have direct path to sdist and wheel files built by tox (to avoid globbing in .tox/dist folder), and this logic should run when tox is done (more specifically it should run after a certain env called "ci" is done, but it's fine if it runs after everything is done)

**Reply by gaborbernat** [reply]
I will rephrase my question then. Is it related to the tox environment or the packaging one? If this is some functionality that makes operation on the package on on the package under test then the answer will be different because you should likely add the functionality into the packaging environment and not the run environment.

**Reply by gaborbernat** [reply]
For the packaging environment, you likely want to subclass the current existing packaging environment and add into the tear down method your custom logic. In the configuration setup hook, you can set your custom packaging class at the tox packaging environment.

**Reply by f3flight** [reply]
Hm good question. My CI custom run env dumps dependency tree (using pipdeptree) into a file and then I collect and process this data via a tox plugin which runs on cleanup. So I need some env where the package is installed and where I can have pipdeptree also installed, to dump dependency tree and export it to our ci/cd.vRight now I have a separate section in tox.ini with an env which installs and runs pipdeptree, and then the plugin collects that (as I could not figure out how to inject a run env programmatically in tox v3, so I did it via tox.ini).
- Does packaging environment install the package?
- can I instruct packaging environment to install an extra package inside itself (pipdeptree) so I could run it on teardown?

**Reply by gaborbernat** [reply]
Sounds like a separate running environment for me that depends on all other running environments. See the depends keyword in configuration. This environment would have pipdeptree installed. That tool can dump the dependency tree for other python environments, not just the one it is installed into, check the --python arg.

**Reply by f3flight** [reply, marked as answer]
Thanks @gaborbernat. I managed to achieve the desired result by using `tox_env_teardown` hook by checking for `tox_env.name` inside (to only run when env is called "ci") and accessing `tox_env._packages[0].path` to get to the sdist. Didn't need the `depends` option as this "ci" env doesn't really depend on others but good to know it exists, I may use it for other needs in future.
